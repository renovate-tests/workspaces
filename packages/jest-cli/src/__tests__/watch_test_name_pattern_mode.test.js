/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

const chalk = require('chalk');
const {KEYS} = require('../constants');

const runJestMock = jest.fn();

let terminalWidth;

jest.mock('ansi-escapes', () => ({
  clearScreen: '[MOCK - clearScreen]',
  cursorDown: (count = 1) => `[MOCK - cursorDown(${count})]`,
  cursorHide: '[MOCK - cursorHide]',
  cursorRestorePosition: '[MOCK - cursorRestorePosition]',
  cursorSavePosition: '[MOCK - cursorSavePosition]',
  cursorShow: '[MOCK - cursorShow]',
  cursorTo: (x, y) => `[MOCK - cursorTo(${x}, ${y})]`,
}));

jest.mock(
  '../search_source',
  () =>
    class {
      findMatchingTests(pattern) {
        return {paths: []};
      }
    },
);

jest.doMock('chalk', () => new chalk.constructor({enabled: false}));

jest.doMock('strip-ansi');
require('strip-ansi').mockImplementation(str => str);

jest.doMock(
  '../run_jest',
  () =>
    function() {
      const args = Array.from(arguments);
      const [{onComplete}] = args;
      runJestMock.apply(null, args);

      // Call the callback
      onComplete({
        snapshot: {},
        testResults: [
          {
            testResults: [{title: 'should return the correct index when'}],
          },
          {
            testResults: [{title: 'should allow test siblings to modify'}],
          },
          {
            testResults: [{title: 'might get confusing'}],
          },
          {
            testResults: [
              {title: 'should handle length properties that cannot'},
            ],
          },
          {
            testResults: [{title: 'should recognize various types'}],
          },
          {
            testResults: [{title: 'should recognize null and undefined'}],
          },
          {
            testResults: [{title: 'should not output colors to pipe'}],
          },
          {
            testResults: [{title: 'should convert string to a RegExp'}],
          },
          {
            testResults: [
              {title: 'should escape and convert string to a RegExp'},
            ],
          },
          {
            testResults: [{title: 'should convert grep string to a RegExp'}],
          },
        ],
      });

      return Promise.resolve();
    },
);

jest.doMock('../lib/terminal_utils', () => ({
  getTerminalWidth: () => terminalWidth,
}));

const watch = require('../watch');

const toHex = char => Number(char.charCodeAt(0)).toString(16);

const globalConfig = {
  watch: true,
};

afterEach(runJestMock.mockReset);

describe('Watch mode flows', () => {
  let pipe;
  let hasteMapInstances;
  let contexts;
  let stdin;

  beforeEach(() => {
    terminalWidth = 80;
    pipe = {write: jest.fn()};
    hasteMapInstances = [{on: () => {}}];
    contexts = [{config: {}}];
    stdin = new MockStdin();
  });

  it('Pressing "T" enters pattern mode', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    // Write a enter pattern mode
    stdin.emit(KEYS.T);
    expect(pipe.write).toBeCalledWith(' pattern â€º ');

    const assertPattern = hex => {
      pipe.write.mockReset();
      stdin.emit(hex);
      expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
    };

    // Write a pattern
    ['c', 'o', 'n', ' ', '1', '2'].map(toHex).forEach(assertPattern);

    [KEYS.BACKSPACE, KEYS.BACKSPACE].forEach(assertPattern);

    ['*'].map(toHex).forEach(assertPattern);

    // Runs Jest again
    runJestMock.mockReset();
    stdin.emit(KEYS.ENTER);
    expect(runJestMock).toBeCalled();

    // globalConfig is updated with the current pattern
    expect(runJestMock.mock.calls[0][0].globalConfig).toMatchObject({
      onlyChanged: false,
      testNamePattern: 'con *',
      watch: true,
      watchAll: false,
    });
  });

  it('can select a specific test name from the typeahead results', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    // Write a enter pattern mode
    stdin.emit(KEYS.T);

    // Write a pattern
    ['c', 'o', 'n']
      .map(toHex)
      .concat([
        KEYS.ARROW_DOWN,
        KEYS.ARROW_DOWN,
        KEYS.ARROW_DOWN,
        KEYS.ARROW_UP,
      ])
      .forEach(key => stdin.emit(key));

    stdin.emit(KEYS.ENTER);

    expect(runJestMock.mock.calls[1][0].globalConfig.testNamePattern).toBe(
      'should convert string to a RegExp',
    );
  });

  it('Results in pattern mode get truncated appropriately', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    stdin.emit(KEYS.T);

    [50, 30].forEach(width => {
      terminalWidth = width;
      stdin.emit(KEYS.BACKSPACE);
      pipe.write.mockReset();
      stdin.emit(KEYS.T);
      expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
    });
  });
});

class MockStdin {
  constructor() {
    this._callbacks = [];
  }

  setRawMode() {}

  resume() {}

  setEncoding() {}

  on(evt, callback) {
    this._callbacks.push(callback);
  }

  emit(key) {
    this._callbacks.forEach(cb => cb(key));
  }
}
