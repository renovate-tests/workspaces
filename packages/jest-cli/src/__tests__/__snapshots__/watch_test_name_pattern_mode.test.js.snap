// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Watch mode flows Pressing "T" enters pattern mode 1`] = `
"


 pattern › c
[MOCK - cursorSavePosition]


 Pattern matches 9 tests from cached test suites

 › should return the correct index when

 › might get confusing

 › should handle length properties that cannot

 › should recognize various types

 › should recognize null and undefined

 › should not output colors to pipe

 › should convert string to a RegExp

 › should escape and convert string to a RegExp

 › should convert grep string to a RegExp
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 2`] = `
"


 pattern › co
[MOCK - cursorSavePosition]


 Pattern matches 8 tests from cached test suites

 › should return the correct index when

 › might get confusing

 › should recognize various types

 › should recognize null and undefined

 › should not output colors to pipe

 › should convert string to a RegExp

 › should escape and convert string to a RegExp

 › should convert grep string to a RegExp
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 3`] = `
"


 pattern › con
[MOCK - cursorSavePosition]


 Pattern matches 4 tests from cached test suites

 › might get confusing

 › should convert string to a RegExp

 › should escape and convert string to a RegExp

 › should convert grep string to a RegExp
[MOCK - cursorTo(14, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 4`] = `
"


 pattern › con 
[MOCK - cursorSavePosition]


 Pattern matches no tests from cached test suites
[MOCK - cursorTo(15, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 5`] = `
"


 pattern › con 1
[MOCK - cursorSavePosition]


 Pattern matches no tests from cached test suites
[MOCK - cursorTo(16, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 6`] = `
"


 pattern › con 12
[MOCK - cursorSavePosition]


 Pattern matches no tests from cached test suites
[MOCK - cursorTo(17, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 7`] = `
"


 pattern › con 1
[MOCK - cursorSavePosition]


 Pattern matches no tests from cached test suites
[MOCK - cursorTo(16, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 8`] = `
"


 pattern › con 
[MOCK - cursorSavePosition]


 Pattern matches no tests from cached test suites
[MOCK - cursorTo(15, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 9`] = `
"


 pattern › con *
[MOCK - cursorSavePosition]


 Pattern matches 4 tests from cached test suites

 › might get confusing

 › should convert string to a RegExp

 › should escape and convert string to a RegExp

 › should convert grep string to a RegExp
[MOCK - cursorTo(16, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Results in pattern mode get truncated appropriately 1`] = `
"


 pattern › t
[MOCK - cursorSavePosition]


 Pattern matches 9 tests from cached test suites

 › should return the correct index when

 › should allow test siblings to modify

 › might get confusing

 › should handle length properties that cannot

 › should recognize various types

 › should not output colors to pipe

 › should convert string to a RegExp

 › should escape and convert string to a RegExp

 › should convert grep string to a RegExp
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Results in pattern mode get truncated appropriately 2`] = `
"


 pattern › t
[MOCK - cursorSavePosition]


 Pattern matches 9 tests from cached test suites

 › should return the corre...

 › should allow test sibli...

 › might get confusing

 › should handle length pr...

 › should recognize variou...

 › should not output color...

 › should convert string t...

 › should escape and conve...

 › should convert grep str...
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]"
`;
